src/main/java/io/reactivex/internal/operators/flowable/FlowableConcatMapEager.java

```diff
@@ -0,0 +1,356 @@
+        this.mapper = mapper;
+        this.maxConcurrency = maxConcurrency;
+        this.prefetch = prefetch;
+        this.errorMode = errorMode;
+    }
+
+    @Override
+    protected void subscribeActual(Subscriber<? super R> s) {
+        source.subscribe(new ConcatMapEagerDelayErrorSubscriber<T, R>(
+                s, mapper, maxConcurrency, prefetch, errorMode));
+    }
+
+    static final class ConcatMapEagerDelayErrorSubscriber<T, R>
+    extends AtomicInteger
+    implements Subscriber<T>, Subscription, InnerQueuedSubscriberSupport<R> {
+        /** */
+        private static final long serialVersionUID = -4255299542215038287L;
+
+        final Subscriber<? super R> actual;
+        
+        final Function<? super T, ? extends Publisher<? extends R>> mapper;
+        
+        final int maxConcurrency;
+        
+        final int prefetch;
+        
+        final ErrorMode errorMode;
+
+        final AtomicReference<Throwable> error;
+        
+        final AtomicLong requested;
+        
+        Subscription s;
+        
+        Queue<InnerQueuedSubscriber<R>> subscribers;
+        
+        volatile boolean cancelled;
+        
+        volatile boolean done;
+        
+        volatile InnerQueuedSubscriber<R> current;
+        
+        public ConcatMapEagerDelayErrorSubscriber(Subscriber<? super R> actual,
+                Function<? super T, ? extends Publisher<? extends R>> mapper, int maxConcurrency, int prefetch,
+                ErrorMode errorMode) {
+            this.actual = actual;
+            this.mapper = mapper;
+            this.maxConcurrency = maxConcurrency;
+            this.prefetch = prefetch;
+            this.errorMode = errorMode;
+            this.subscribers = new SpscLinkedArrayQueue<InnerQueuedSubscriber<R>>(Math.min(prefetch, maxConcurrency));
+            this.error = new AtomicReference<Throwable>();
+            this.requested = new AtomicLong();
+        }
+        
+        @Override
+        public void onSubscribe(Subscription s) {
+            if (SubscriptionHelper.validate(this.s, s)) {
+                this.s = s;
+                
+                actual.onSubscribe(this);
+                
+                s.request(maxConcurrency == Integer.MAX_VALUE ? Long.MAX_VALUE : maxConcurrency);
+            }
+            
+        }
+        
+        @Override
+        public void onNext(T t) {
+            Publisher<? extends R> p;
+            
+            try {
+                p = Objects.requireNonNull(mapper.apply(t), "The mapper returned a null Publisher");
+            } catch (Throwable ex) {
+                Exceptions.throwIfFatal(ex);
+                s.cancel();
+                onError(ex);
+                return;
+            }
+            
+            InnerQueuedSubscriber<R> inner = new InnerQueuedSubscriber<R>(this, prefetch);
+            
+            if (!cancelled) {
+                subscribers.offer(inner);
+                if (!cancelled) {
```

Comment:
nit: can we invert that and make `drainAndCancel` first so it'll be easier to read and harder to make some mistake during refactoring?